fintech-examples/fintech-impl/src/main/java/de/adorsys/opba/fintech/impl/service/RedirectHandlerService.java

```diff
@@ -0,0 +1,94 @@
+    private final AuthorizeService authorizeService;
+
+    public RedirectUrlsEntity registerRedirectUrlForSession(String xsrfToken) {
+        String redirectCode = UUID.randomUUID().toString();
+        log.debug("ONLY FOR DEBUG: redirectCode: {}", redirectCode);
+
+        RedirectUrlsEntity redirectUrls = new RedirectUrlsEntity();
+
+        redirectUrls.setRedirectCode(redirectCode);
+        redirectUrls.setRedirectState(xsrfToken);
+        redirectUrls.setNotOkURL(getModifiedUrlWithRedirectCode(notOkUrl, redirectCode));
+        redirectUrls.setOkURL(getModifiedUrlWithRedirectCode(okUrl, redirectCode));
+
+        return redirectUrlRepository.save(redirectUrls);
+    }
+
+    public ResponseEntity doRedirect(String redirectState, String redirectId, String redirectCode) {
+        if (StringUtils.isBlank(redirectCode)) {
+            log.warn("Validation redirect request was failed: redirect code is empty!");
+            return prepareRedirectResponse(exceptionUrl);
+        }
+
+        Optional<RedirectUrlsEntity> redirectUrlsEntityOptional = redirectUrlRepository.findByRedirectCode(redirectCode);
+
+        if (!redirectUrlsEntityOptional.isPresent()) {
+            log.warn("Validation redirect request was failed: redirect code is wrong!");
+            return prepareRedirectResponse(exceptionUrl);
+        }
+
+        RedirectUrlsEntity redirectUrlsEntity = redirectUrlsEntityOptional.get();
+
+        if (StringUtils.isBlank(redirectState)) {
+            log.warn("Validation redirect request was failed: Xsrf Token is empty!");
+            return prepareRedirectResponse(redirectUrlsEntity.getNotOkURL());
+        }
+
+        if (!authorizeService.isAuthorized(redirectState, null)) {
+            log.warn("Validation redirect request was failed: Xsrf Token is wrong or user are not authorized!");
+            return prepareRedirectResponse(redirectUrlsEntity.getNotOkURL());
+        }
+
+        SessionEntity optionalUser = authorizeService.getByXsrfToken(redirectState);
+        updateSessionByRedirectCode(optionalUser, redirectCode);
+
+        return prepareRedirectResponse(redirectUrlsEntity.getOkURL());
+    }
+
+    private void updateSessionByRedirectCode(SessionEntity sessionEntity, String redirectCode) {
+        PsuConsentSessionResponse psuConsentSessionResponse = new TppBankingApiTokenMock().getTransactionsResponse(redirectCode);
+        sessionEntity.setPsuConsentSession(psuConsentSessionResponse.getPsuConsentSession().toString());
+        authorizeService.updateUserSession(sessionEntity);
+    }
+
+    private ResponseEntity prepareRedirectResponse(String redirectUrl) {
+        MultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
+        headers.put(LOCATION_HEADER, singletonList(redirectUrl));
+
+        return new ResponseEntity<>(headers, HttpStatus.FOUND);
+    }
+
+    private String getModifiedUrlWithRedirectCode(String url, String redirectCode) {
```

Comment:
Can't we use templates in URL and expand them accordingly?
like http://google.com/${redirectCode}

Reply:
done